torch/package/_mangling.py

```diff
@@ -0,0 +1,140 @@
+results are never saved into a package. Assigning mangle parents on import
+means that we can enforce that mangle parents are unique within the
+environment doing the importing. It also allows us to avoid serializing (and
+maintaining backward compatibility for) this detail.
+
+.. note::
+    No mangle parents should ever be saved into a package.
+
+Mangled names in PackageExporter
+--------------------------------
+Occasionally ``PackageExporter`` may encounter mangled names during export. For
+example, the user may be re-packaging an object that was imported (and thus
+had its module name mangled).
+
+This means that ``PackageExporter`` must demangle all module names before it
+writes them to the package. This is needed in two places:
+    - When converting the module name to a filename to for placing source code in the package.
+    - When saving the module name in the pickle GLOBAL opcode.
+"""
+import re
+
+_mangle_index = 0
+
+
+class PackageMangler:
+    """
+    Used on import, to ensure that all modules imported have a shared mangle parent.
+    """
+
+    def __init__(self):
+        global _mangle_index
+        self._mangle_index = _mangle_index
+        # Increment the global index
+        _mangle_index += 1
+        self._mangle_parent = f"__torch_package_{self._mangle_index}__"
+
+    def mangle(self, name):
+        return self._mangle_parent + "." + name
+
+    def demangle(self, mangled):
+        """
+        Note: This only demangles names that were mangled by this specific
+        PackageMangler. It will pass through names created by a different
+        PackageMangler instance.
+        """
+        if mangled.startswith(self._mangle_parent + "."):
+            return mangled.partition(".")[2]
+
+        # wasn't a mangled name
+        return mangled
+
+    def parent_name(self):
+        return self._mangle_parent
+
+
+class PackageDemangler:
+    """
+    Used on export, to ensure that we are only writing demangled names to the pacakge.
+    """
+
+    def __init__(self):
+        # Map of demangled name => original name
+        # This is to detect name collisions when you are trying to save something like:
+        #    foo.bar => foo.bar
+        #    __torch0__.foo.bar => foo.bar   # collision!
+        #    __torch1__.foo.bar => foo.bar   # collision!
+        self.demangled_names: Dict[str, str] = {}
+
+    def demangle(self, name):
+        """
+        Note: Unlike PackageMangler.demangle, this version work on any
+        mangled name, irrespective of which PackageMangler created it.
+        """
+        if not _was_mangled(name):
+            self.demangled_names[name] = name
+            return name
+
+        demangled = name.partition(".")[2]
+        # Check for whether this demangled name collides with a previously saved module.
+        existing_name = self.demangled_names.get(demangled)
+        if existing_name is None:
+            self.demangled_names[demangled] = name
```

Comment:
Putting state into the demangler is surprising. I can imagine situations where someone doesn't realize this object is stateful and then makes a copy of it, breaking the checks. It would be better for this logic to be handled in the export process which is already making sure a package is only exported the first time it is seen.